hir/src/asm/builder.rs

```diff
@@ -1345,3 +1420,327 @@ impl<'a> MasmOpBuilder<'a> {
+/// lb.ins().mod_imm_u32(4); // [element_offset, array + word_offset, ..]
+///
+/// // Precalculate what elements of the word to drop, so that
+/// // we are only left with the specific element we wanted
+/// lb.ins().dup(0);     // [element_offset, element_offset, ..]
+/// lb.ins().lt_imm(Felt::new(3)); // [element_offset < 3, element_offset, ..]
+/// lb.ins().dup(1);     // [element_offset, element_offset < 3, ..]
+/// lb.ins().lt_imm(Felt::new(2)); // [element_offset < 2, element_offset < 3, ..]
+/// lb.ins().dup(2);     // [element_offset, element_offset < 2, ..]
+/// lb.ins().lt_imm(Felt::new(1)); // [element_offset < 1, element_offset < 2, ..]
+///
+/// // Load the word
+/// lb.ins().dup(4);      // [array + word_offset, element_offset < 1]
+/// lb.ins().loadw(); // [word[0], word[1], word[2], word[3], element_offset < 1]
+///
+/// // Select the element, `E`, that we want by conditionally dropping
+/// // elements on the operand stack with a carefully chosen sequence
+/// // of conditionals: E < N forall N in 0..=3
+/// lb.ins().movup(4);   // [element_offset < 1, word[0], ..]
+/// lb.ins().cdrop();    // [word[0 or 1], word[2], word[3], element_offset < 2]
+/// lb.ins().movup(3);   // [element_offset < 2, word[0 or 1], ..]
+/// lb.ins().cdrop();    // [word[0 or 1 or 2], word[3], element_offset < 3]
+/// lb.ins().movup(2);   // [element_offset < 3, ..]
+/// lb.ins().cdrop();    // [array[i], i, len, sum, array]
+/// lb.ins().movup(3);   // [sum, array[i], i, len, array]
+/// lb.ins().add();      // [sum + array[i], i, len, array]
+/// lb.ins().movdn(2);   // [i, len, sum + array[i], array]
+///
+/// // We've reached the end of the loop, but we need a copy of the
+/// // loop header here in order to use the expression `i < len` as
+/// // the condition for the loop
+/// lb.ins().dup(0);     // [i, i, len, ..]
+/// lb.ins().dup(2);     // [len, i, i, len, ..]
+/// lb.ins().lt();       // [i < len, i, len, sum, array]
+///
+/// // Finalize, it is at this point that validation will occur
+/// lb.build();
+/// ```
+pub struct LoopBuilder<'a> {
+    dfg: &'a mut DataFlowGraph,
+    asm: &'a mut InlineAsm,
+    /// This reference is to the operand stack in the parent [MasmOpBuilder],
+    /// which represents the operand stack on entry to the loop. Upon finalization
+    /// of the loop, we use update this operand stack to reflect the state upon
+    /// exit from the loop.
+    ///
+    /// In effect, when the call to `while_true` or `repeat` returns, the operand
+    /// stack in the parent builder will look as if the loop instruction has finished
+    /// executing.
+    in_stack: &'a mut OperandStack<Type>,
+    /// This is the operand stack state within the loop.
+    ///
+    /// Upon finalization of the loop instruction, this state is used to validate
+    /// the effect of the loop body on the operand stack. For `repeat`, which is
+    /// unconditionally entered, no special validation is performed. However, for
```

Comment:
Shouldn't `repeat` loops be validated as well? The stack could grow with each iteration, no?

Reply:
`repeat.n` can't produce dynamic (i.e. not statically analyzable) changes to the operand stack, it is statically expanded to `n` copies of the `repeat` body. In other words, if you make a modification to the stack, we know exactly how many times that modification will occur, so we can model the state of the stack across every iteration, and afterwards.

That said, I agree that some additional validation is needed here:

* We want to ensure that, given _S_ elements on the operand stack upon entry to the loop, that the top _S_ elements on the stack along the loopback edge match those that were on the stack on entry, for all `n` iterations.
* We want to ensure that if each iteration removes elements from the stack, that all `n` iterations can be performed without performing an invalid stack operation.

I'll do that before we merge this.